# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/ipc/RemoteVideoDecoder.cpp
# Commit: 9e0809ed54cd
# Full Hash: 9e0809ed54cd2da0e3ebbb0bb143f85bb6ad6b14
# Author: Dan Glastonbury <dan.glastonbury@gmail.com>
# Date: 2020-10-21 09:51:28
# Regressor Bug: 1595994
# File Overlap Count: 1
# Description:
#   Bug 1595994 - P4: Use PDMFactory to create decoders in RDD & GPU process. r=kamidphish
#   
#   Depends on D54876
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54875
# ==============================================================================

diff -r e3f920da2232 -r 9e0809ed54cd dom/media/ipc/RemoteVideoDecoder.cpp
--- a/dom/media/ipc/RemoteVideoDecoder.cpp	Tue Oct 20 23:24:27 2020 +0000
+++ b/dom/media/ipc/RemoteVideoDecoder.cpp	Tue Oct 20 23:24:40 2020 +0000
@@ -159,47 +159,22 @@
         KnowsCompositorVideo::TryCreateForIdentifier(*aIdentifier);
   }
 
-  RefPtr<layers::ImageContainer> container = new layers::ImageContainer();
+  // It is possible for CreateDecoder() below to fail. In that case, we need to
+  // free the ImageContainer to avoid it leaking.
+  auto imageContainer = MakeRefPtr<layers::ImageContainer>();
   if (mKnowsCompositor && XRE_IsRDDProcess()) {
     // Ensure to allocate recycle allocator
-    container->EnsureRecycleAllocatorForRDD(mKnowsCompositor);
+    imageContainer->EnsureRecycleAllocatorForRDD(mKnowsCompositor);
   }
-
-  CreateDecoderParams params(mVideoInfo);
-  params.mKnowsCompositor = mKnowsCompositor;
-  params.mImageContainer = container;
-  params.mRate = CreateDecoderParams::VideoFrameRate(aFramerate);
-  params.mOptions = aOptions;
   MediaResult error(NS_OK);
-  params.mError = &error;
-
-  RefPtr<MediaDataDecoder> decoder;
-  if (XRE_IsGPUProcess()) {
-#ifdef XP_WIN
-    // Ensure everything is properly initialized on the right thread.
-    PDMFactory::EnsureInit();
+  auto params = CreateDecoderParams{
+      mVideoInfo,     mKnowsCompositor,
+      imageContainer, CreateDecoderParams::VideoFrameRate(aFramerate),
+      aOptions,       CreateDecoderParams::NoWrapper(true),
+      &error};
 
-    // TODO: Ideally we wouldn't hardcode the WMF PDM, and we'd use the normal
-    // PDM factory logic for picking a decoder.
-    RefPtr<WMFDecoderModule> pdm(new WMFDecoderModule());
-    pdm->Startup();
-    decoder = pdm->CreateVideoDecoder(params);
-#else
-    MOZ_ASSERT(false,
-               "Can't use RemoteVideoDecoder in the GPU process on non-Windows "
-               "platforms yet");
-#endif
-  }
-
-#ifdef MOZ_AV1
-  if (AOMDecoder::IsAV1(params.mConfig.mMimeType)) {
-    if (StaticPrefs::media_av1_use_dav1d()) {
-      decoder = new DAV1DDecoder(params);
-    } else {
-      decoder = new AOMDecoder(params);
-    }
-  }
-#endif
+  auto& factory = aParent->EnsurePDMFactory();
+  RefPtr<MediaDataDecoder> decoder = factory.CreateDecoder(params);
 
   if (NS_FAILED(error)) {
     MOZ_ASSERT(aErrorDescription);